### DISCUSSION TITLE
Why does only xStream raise alerts in my streaming loop?

### DISCUSSION THREAD

**Discussion by selimfirat** [selimfirat/pysad#64] [Q&A]
*Originally asked in #5.*

**Q:** I copied `example_usage.py`, added a `GaussianTailProbabilityCalibrator`, and swapped models. Only xStream ever raises an alert. The others either fail or never flag anything. Why?

**A:** The loop called `calibrator.fit_transform(score)`. The non-`_partial` methods expect an array with one score per instance. Many models return a plain float, so that call raises an error, and the bare `except:` quietly set every probability to 0. xStream happens to return a one-element array, so it slipped through.

For one score at a time, use the `_partial` methods:

```python
import numpy as np
from pysad.models import RSHash
from pysad.transform.probability_calibration import GaussianTailProbabilityCalibrator

X = np.random.rand(500, 3)
model = RSHash(feature_mins=np.zeros(3), feature_maxes=np.ones(3))
calibrator = GaussianTailProbabilityCalibrator(window_size=100)

for i, x in enumerate(X):
    score = model.fit_score_partial(x)              # one score per instance
    prob = calibrator.fit_transform_partial(score)  # not calibrator.fit_transform(score)
    if prob > 0.99:
        print(f"Alert at {i}: p={prob:.3f}")
```

Also worth checking:

- Avoid bare `except:` while you experiment, so errors show up.
- Look at AUROC before tuning alert thresholds.
- `RunningAveragePostprocessor` can smooth out short spikes.
- Hyperparameters matter. For example, `ExactStorm`'s default `max_radius=0.1` gives constant scores on unit-normalized arrhythmia data.
